src/boot/bootctl-status.c

```diff
@@ -503,6 +504,224 @@ int verb_status(int argc, char *argv[], void *userdata) {
+
+        r = ref_file(known_files, fn, -1);
+        if (r == 0) {
+                if (arg_dry)
+                        r = chase_symlinks_and_access(fn, root, CHASE_PREFIX_ROOT|CHASE_PROHIBIT_SYMLINKS, F_OK, &path, NULL);
+                else
+                        r = chase_symlinks_and_unlink(fn, root, CHASE_PREFIX_ROOT|CHASE_PROHIBIT_SYMLINKS, 0, &path);
+                if (r < 0) {
+                        log_error_errno(errno, "Failed to remove \"%s\": %m", path);
+                        return;
+                }
+                log_info("%s %s", arg_dry?"Would remove":"Removed", path);
+        }
+}
+
+static int count_known_files(const BootConfig *config, const char* root, Hashmap **ret_known_files) {
+        _cleanup_(hashmap_free_free_keyp) Hashmap *known_files = NULL;
+        int r = 0;
+        assert(config);
+        assert(ret_known_files);
+
+        known_files = hashmap_new(&path_hash_ops);
+        if (!known_files)
+                return log_oom();
+
+        for (size_t i = 0; i < config->n_entries; i++) {
+                const BootEntry *e = config->entries + i;
+
+                if (!e->root || strcmp(e->root, root))
+                        continue;
+
+                r = ref_file(known_files, e->kernel, 1);
+                if (r < 0)
+                        return r;
+                STRV_FOREACH(s, e->initrd) {
+                        r = ref_file(known_files, *s, 1);
+                        if (r < 0)
+                                return r;
+                }
+                r = ref_file(known_files, e->device_tree, 1);
+                if (r < 0)
+                        return r;
+                STRV_FOREACH(s, e->device_tree_overlay) {
+                        r = ref_file(known_files, *s, 1);
+                        if (r < 0)
+                                return r;
+                }
+        }
+
+        *ret_known_files = TAKE_PTR(known_files);
+
+        return 0;
+}
+
+static int unlink_entry(const BootConfig *config, const char *root, const char *fn) {
+        _cleanup_(hashmap_free_free_keyp) Hashmap *known_files = NULL;
+        const BootEntry *e = NULL;
+        int r;
+
+        r = count_known_files(config, root, &known_files);
+        if (r < 0)
+                return log_error_errno(r, "Failed to count files in %s: %m", root);
+
+        for (size_t i = 0; i < config->n_entries; i++) {
+                const BootEntry *ei = &config->entries[i];
+
+                if (!ei->root || strcmp(ei->root, root))
+                        continue;
+                if (fnmatch(fn, ei->id, FNM_CASEFOLD))
+                        continue;
+
+                e = ei;
+                if (i == (size_t) config->default_entry)
+                        log_warning("%s is the default boot entry", fn);
+                if (i == (size_t) config->selected_entry)
+                        log_warning("%s is the selected boot entry", fn);
+                break;
+        }
```

Comment:
btw could you split this part out, and make it a helper of its own? maybe "boot_config_find_pattern()" or so

Reply:
(and do the default/selected check outside of the function

Reply:
btw, we probably should refuse to unlink the default/selected entry by default, for safety reasons, no?

and then add a new `--force` switch or so which can override this and remove it anyway?

Reply:
> btw could you split this part out, and make it a helper of its own? maybe "boot_config_find_pattern()" or so

There's `boot_config_find()` already but it's private to `src/shared/bootspec.c`. Could either duplicate locally or export.

Reply:
> btw, we probably should refuse to unlink the default/selected entry by default, for safety reasons, no?
> 
> and then add a new `--force` switch or so which can override this and remove it anyway?

The idea is to call `bootctl unlink` from `kernel-install` which in turn gets called eg by package management when removing the kernel. So that one would have to pass `--force` always then. Ie no gain on that front at least.

Reply:
> > btw could you split this part out, and make it a helper of its own? maybe "boot_config_find_pattern()" or so
> 
> There's `boot_config_find()` already but it's private to `src/shared/bootspec.c`. Could either duplicate locally or export.

ah well, ok, that one doesn't check for same root.